pymess/models/common.py

```diff
@@ -18,3 +28,91 @@ def create_from_related_objects(self, *related_objects):
+
+    STATE = ChoicesNumEnum()
+
+    sent_at = models.DateTimeField(verbose_name=_('sent at'), null=True, blank=True, editable=False)
+    recipient = models.CharField(verbose_name=_('recipient'), null=False, blank=False, max_length=20)
+    content = models.TextField(verbose_name=_('content'), null=False, blank=False)
+    template_slug = models.SlugField(verbose_name=_('slug'), max_length=100, null=True, blank=True, editable=False)
+    state = models.IntegerField(verbose_name=_('state'), null=False, blank=False, choices=STATE.choices, editable=False)
+    backend = models.CharField(verbose_name=_('backend'), null=True, blank=True, editable=False, max_length=250)
+    error = models.TextField(verbose_name=_('error'), null=True, blank=True, editable=False)
+    extra_data = JSONField(verbose_name=_('extra data'), null=True, blank=True, editable=False)
+    extra_sender_data = JSONField(verbose_name=_('extra sender data'), null=True, blank=True, editable=False)
+    tag = models.SlugField(verbose_name=_('tag'), null=True, blank=True, editable=False)
+
+    def __str__(self):
+        return self.recipient
+
+    class Meta:
+        abstract = True
+        ordering = ('-created_at',)
+
+
+class BaseRelatedObject(SmartModel):
+
+    content_type = models.ForeignKey(ContentType, verbose_name=_('content type of the related object'),
+                                     null=False, blank=False, on_delete=models.CASCADE)
+    object_id = models.TextField(verbose_name=_('ID of the related object'), null=False, blank=False)
+    object_id_int = models.PositiveIntegerField(verbose_name=_('ID of the related object in int format'), null=True,
+                                                blank=True, db_index=True)
+    content_object = GenericForeignKey('content_type', 'object_id')
+
+    objects = RelatedObjectManager()
+
+    class Meta:
+        abstract = True
+        ordering = ('-created_at',)
+
+
+class BaseAbstractTemplate(SmartModel):
+
+    slug = models.SlugField(verbose_name=_('slug'), max_length=100, null=False, blank=False, editable=False,
+                            primary_key=True)
+    body = models.TextField(verbose_name=_('message body'), null=True, blank=False)
+
+    def _update_context_data(self, context_data):
+        return context_data
+
+    def render_body(self, context_data):
+        context_data = self._update_context_data(context_data)
+        return Template(self.get_body()).render(Context(context_data))
+
+    def clean_body(self, context_data=None):
+        try:
+            self.render_body(context_data or {})
+        except (TemplateSyntaxError, TemplateDoesNotExist) as ex:
+            raise ValidationError(_('Error during template body rendering: "{}"').format(ex))
+
+    def get_body(self):
+        return self.body
+
+    def can_send(self, recipient, context_data):
+        return True
+
+    def get_backend_sender(self):
+        raise NotImplemented('implement in subclass')
```

Comment:
Tady má být `NotImplementedError`.

Reply:
OK.

je to tam aj-aj, ale teda toto som dal bez. asi fakt nebude nutna sprava...